### PR TITLE
feat(lib): Introduce TerraformDynamicBlock to enable iterator usage on blocks

### LINKED ISSUE #1798: Implement block-level iterators
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

To support dynamic list iterations on the block / resource level we will introduce a new API, called Iterator.

In this issue we are building the block-level Iterator API. To create dynamic blocks one can use iterator.dynamic([...]) and the block will be repeated for each item in the list.

```ts
import { TerraformStack, Iterator } from "cdktf";
import { S3Bucket, DataAwsSubnets, DataAwsAllowedOrigins } from "./.gen/providers/aws-ficticious";
import { Construct } from "constructs";

export class MyStack extends TerraformStack {
 constructor(scope: Construct, id: string) {
   super(scope, id);

  // Fictitious data source that returns [{method, origin}]
   const allowedOrigins = new DataAwsAllowedOrigins(
     this,
     "possibleOrigins",
     {}
   );
   const allowedOriginsIterator = Iterator.fromList(allowedOrigins.origins);
   new S3Bucket(this, "my-bucket", {
     corsRule: allowedOriginsIterator.dynamic(
       {
         allowedMethods: [allowedOriginsIterator.getStringProperty("method")],
         allowedOrigins: [allowedOriginsIterator.getStringProperty("origin")],
         maxAgeSeconds: 42,
       },
     ),
   });
 }
}
```

### References

- Part of #1094
- Depends on #1797

### PR DESCRIPTION
**Pull request by ansgarm**
Closes #1798

### Open tasks
- [x] Add unit test for error case if TerraformDynamicBlock is used at the wrong position and tried to be resolved (should throw an error; e.g. could happen if passed to a Terraform function)
- [x] How to handle normal lists (non-block type lists)? Convert to some other expression (using Terraform functions) ~~or at least throw an Error if used at an invalid location. Not sure how to detect this yet.~~
- [x] Add edge provider tests for iteration in the various languages
- [x] Add tests that invoke at least `terraform validate` to confirm result is valid (might be something to add to edge provider tests)
- [x] ~~Add docs to iterator docs with an example for block level iteration (in each language)~~ will follow in a separate PR as they will be based on a branch that is not merged yet
- [x] make sure there are follow-up issues for all new `todo:` comments in the code base

### CODE REVIEW

File: packages/cdktf/lib/runtime.ts [hunk 1]
```diff
@@ -25,12 +27,31 @@ export const booleanToTerraform: Mapper = identity;
...
       return x;
     }
 
+    // replace dynamic expressions for block types so they can be detected and replaced properly by processDynamicAttributes()
+    // which also relocates them to dynamic.attributeName (something we can't do when resolving a value, put it somewhere else)
+    if (
+      TerraformDynamicExpression.isTerraformDynamicExpression(x) &&
+      isBlockType
+    ) {
+      return TerraformDynamicBlock.fromDynamicExpression(x);
```
Comment (DanielMSchmidt) [inline packages/cdktf/lib/runtime.ts:52]: I like this wording!

File: packages/cdktf/lib/runtime.ts [hunk 2]
```diff
@@ -25,12 +27,31 @@ export const booleanToTerraform: Mapper = identity;
...
   return (x: any) => {
     if (!canInspect(x)) {
       return x;
     }
 
+    // replace dynamic expressions for block types so they can be detected and replaced properly by processDynamicAttributes()
+    // which also relocates them to dynamic.attributeName (something we can't do when resolving a value, put it somewhere else)
+    if (
+      TerraformDynamicExpression.isTerraformDynamicExpression(x) &&
+      isBlockType
```
Comment (DanielMSchmidt) [inline packages/cdktf/lib/runtime.ts:50]: Can we link / mention here how the `!isBlockType` is handled / resolved?

File: packages/cdktf/lib/terraform-dynamic-block.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import { captureStackTrace } from "./tokens/private/stack-trace";
+
+const DYNAMIC_BLOCK_SYMBOL = Symbol.for("cdktf/TerraformDynamicBlock");
+
+export class TerraformDynamicBlock implements IResolvable {
+  public readonly creationStack: string[];
+  public readonly forEach: ITerraformIterator;
+  public readonly iterator = "each"; // name of temporary variable used in iteration, needed in the future when allowing nesting iterators
+  public readonly content: { [key: string]: any };
+  // labels: TODO: support labels, but they seem to be quite rare (see https://discuss.hashicorp.com/t/labels-in-dynamic-block/21461/5)
```
Comment (DanielMSchmidt) [inline packages/cdktf/lib/terraform-dynamic-block.ts:13]: We should make a follow up issue to support labels

File: packages/cdktf/lib/tfExpression.ts [hunk 4]
```diff
@@ -374,6 +374,72 @@ export function call(name: string, args: Expression[]) {
...
+  }
+
+  public resolve(context: IResolveContext): string {
+    const suppressBraces = context.suppressBraces;
+    context.suppressBraces = true;
+    const key = this.resolveArg(context, FOR_EXPRESSION_KEY);
+    const value = this.resolveArg(context, FOR_EXPRESSION_VALUE);
+    const input = this.resolveArg(context, this.input);
+    const valueExpr = this.resolveArg(context, this.valueExpression);
+    const keyExpr = this.keyExpression
```
Comment (DanielMSchmidt) [inline packages/cdktf/lib/tfExpression.ts:399]: Nit: I'd inline this var to remove a layer of indirection

File: packages/cdktf/lib/tfExpression.ts [hunk 5]
```diff
@@ -374,6 +374,72 @@ export function call(name: string, args: Expression[]) {
...
+}
+
+/**
+ * Creates a for expression, used for advanced internal use cases (e.g. in the @cdktf/aws-adapter)
+ * It is currently not possible to use this via JSII in other languages. Usually you might want to
+ * use TerraformIterator instead.
+ *
+ * @param input the input for the expression e.g. a list or an object
+ * @param valueExpression will express the value e.g. [ for i in &lt;input&gt; : &lt;valueExpression&gt; ]
+ * @param keyExpression if set will return an object e.g. { for k,v in &lt;input&gt; : &lt;keyExpression&gt; => &lt;valueExpression&gt;}
```
Comment (DanielMSchmidt) [inline packages/cdktf/lib/tfExpression.ts:421]: Is the `&lt;` and `&gt;` by design or accident?

File: packages/cdktf/lib/util.ts [hunk 6]
```diff
@@ -96,3 +97,32 @@ export function keysToSnakeCase(object: any): any {
...
   }, {});
 }
+
+/**
+ * dynamic attributes are located at a different position than normal block attributes
+ * This method detects them and moves them from .attributeName to .dynamic.attributeName
+ * It also invokes the .toTerraform() method on the dynamic attribute to get the correct
+ * Terraform representation
+ *
+ * TODO: this does not move nested blocks (do they exist at all?)
```
Comment (DanielMSchmidt) [inline packages/cdktf/lib/util.ts:107]: Can we detect them and throw?

File: packages/cdktf/test/helper/data-source.ts [hunk 7]
```diff
@@ -8,17 +8,20 @@ import {
...
 
 export interface TestDataSourceConfig extends TerraformMetaArguments {
   name: string;
+  listBlock?: IResolvable;
 }
 
 export class TestDataSource extends TerraformDataSource {
   public static readonly tfResourceType: string = "test_data_source";
   public name: string;
+  public listBlock?: IResolvable; // TODO: add special type as well?
```
Comment (DanielMSchmidt) [inline packages/cdktf/test/helper/data-source.ts:24]: Todo left, What special type?

File: packages/cdktf/test/helper/resource.ts [hunk 8]
```diff
@@ -18,6 +23,7 @@ export class TestResource extends TerraformResource {
   public names?: string[];
   public tags?: { [key: string]: string };
   public nestedType?: { [key: string]: string };
+  public listBlock?: IResolvable; // TODO: add special type?
```
Comment (DanielMSchmidt) [inline packages/cdktf/test/helper/resource.ts:26]: TODO left

File: test/typescript/iterators/main.ts [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    });
+
+    const it = TerraformIterator.fromList(list.listValue);
+
+    const files = new TerraformAsset(this, "files-dir", {
+      path: "./files",
+      type: AssetType.DIRECTORY,
+    });
+
+    // TODO: we don't support using .dynamic for simple mapping (i.e. to a string instead of to an object) yet
```
Comment (DanielMSchmidt) [inline test/typescript/iterators/main.ts:34]: Do we throw in these cases?

File: test/typescript/iterators/main.ts [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    new ArchiveProvider.DataArchiveFile(this, "dir", {
+      type: "zip",
+      outputPath: "${path.module}/out1.zip",
+      sourceDir: files.path,
+      // excludes: it.dynamic(`${it.value}.txt`), // allow this at some point
+    });
+
+    // HCL equivalent:
```
Comment (DanielMSchmidt) [inline test/typescript/iterators/main.ts:42]: Love it!

File: test/typescript/iterators/main.ts [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+            },
+          }),
+        },
+      },
+    });
+  }
+}
+
+const app = Testing.stubVersion(new App({}));
+new TestIterators(app, "test-iterators");
```
Comment (DanielMSchmidt) [inline test/typescript/iterators/main.ts:170]: Nice separation!

**Review by DanielMSchmidt — COMMENTED** [review]
Looks good, great work! Just a couple of comments :)

File: packages/cdktf/lib/runtime.ts (hunk 2 above)
Comment (ansgarm) [reply packages/cdktf/lib/runtime.ts:50]: Nice idea! Added a comment

File: packages/cdktf/lib/tfExpression.ts (hunk 5 above)
Comment (ansgarm) [reply packages/cdktf/lib/tfExpression.ts:421]: by design, if you hover over the method name in e.g. VSCode you can see that they look right this way (else they were treated as html tags)

File: packages/cdktf/lib/tfExpression.ts (hunk 4 above)
Comment (ansgarm) [reply packages/cdktf/lib/tfExpression.ts:399]: Am I missing something or did you miss the next two lines? 😅

File: packages/cdktf/test/helper/data-source.ts (hunk 7 above)
Comment (ansgarm) [reply packages/cdktf/test/helper/data-source.ts:24]: Just for testing. Real life bindings don't have an IResolvable here but a type with an interface as well. I think keeping it simple for now is okay as we currently only use the IResolvable case in tests that use this TestDataSource (for the interfaces we use the edge provider tests).

Comment (ansgarm) [reply packages/cdktf/test/helper/data-source.ts:24]: reworded the comment to reflect what I just said

File: packages/cdktf/test/helper/resource.ts (hunk 8 above)
Comment (ansgarm) [reply packages/cdktf/test/helper/resource.ts:26]: changed to a better wording, same as in the data source

File: test/typescript/iterators/main.ts (hunk 9 above)
Comment (ansgarm) [reply test/typescript/iterators/main.ts:34]: No, currently only our types would prohibit this. And I don't think we need to throw, as it would already work to pass a string or number from a technical perspective. It just is not intended yet and not tested.

File: packages/cdktf/lib/util.ts (hunk 6 above)
Comment (ansgarm) [reply packages/cdktf/lib/util.ts:107]: no need for that, as this was an obsolete todo (we now support nested blocks)

File: packages/cdktf/lib/terraform-dynamic-block.ts (hunk 3 above)
Comment (ansgarm) [reply packages/cdktf/lib/terraform-dynamic-block.ts:13]: done

File: test/typescript/iterators/main.ts (hunk 9 above)
Comment (ansgarm) [reply test/typescript/iterators/main.ts:34]: created and linked

**Review by DanielMSchmidt — APPROVED** [review]
Besides from the missing docs this looks great 👍
